grpc/src/main/java/com/linecorp/armeria/server/grpc/DefaultUnframedGrpcErrorHandler.java

```diff
@@ -0,0 +1,289 @@
+                                                                           grpcCode.value())
+                                                                   .build();
+            final HttpData content;
+            try (TemporaryThreadLocals ttl = TemporaryThreadLocals.acquire()) {
+                final StringBuilder msg = ttl.stringBuilder();
+                msg.append("grpc-code: ").append(grpcCode.name());
+                if (grpcMessage != null) {
+                    msg.append(", ").append(grpcMessage);
+                }
+                if (cause != null && ctx.config().verboseResponses()) {
+                    msg.append("\nstack-trace:\n").append(Exceptions.traceText(cause));
+                }
+                content = HttpData.ofUtf8(msg);
+            }
+            return HttpResponse.of(responseHeaders, content);
+        };
+    }
+
+    /**
+     * Returns a rich error JSON response based on Google APIs.
+     * Please refer to <a href="https://cloud.google.com/apis/design/errors#error_model">Google error model</a>
+     * @param statusMappingFunction The function which maps the {@link Throwable} or gRPC {@link Status} code
+     *                              to an {@link HttpStatus} code.
+     */
+    static UnframedGrpcErrorHandler ofRichJson(UnframedGrpcStatusMappingFunction statusMappingFunction) {
+        final UnframedGrpcStatusMappingFunction mappingFunction =
+                requireNonNull(statusMappingFunction, "statusMappingFunction")
+                        .orElse(UnframedGrpcStatusMappingFunction.of());
+        return (ctx, status, response) -> {
+            final ByteBuf buffer = ctx.alloc().buffer();
+            final Code grpcCode = status.getCode();
+            final String grpcMessage = status.getDescription();
+            final Throwable cause = getThrowableFromContext(ctx);
+            final HttpStatus httpStatus = mappingFunction.apply(ctx, status, cause);
+            final HttpHeaders trailers = !response.trailers().isEmpty() ?
+                                         response.trailers() : response.headers();
+            final String grpcStatusDetailsBin = trailers.get(GrpcHeaderNames.GRPC_STATUS_DETAILS_BIN);
+            final ResponseHeaders responseHeaders = ResponseHeaders.builder(httpStatus)
+                                                                   .contentType(MediaType.JSON_UTF_8)
+                                                                   .addInt(GrpcHeaderNames.GRPC_STATUS,
+                                                                           grpcCode.value())
+                                                                   .build();
+            boolean success = false;
+            try (OutputStream outputStream = new ByteBufOutputStream(buffer);
+                 JsonGenerator jsonGenerator = mapper.createGenerator(outputStream)) {
+                jsonGenerator.writeStartObject();
+                jsonGenerator.writeFieldName("error");
```

Comment:
Sorry, one last question.
Should we just return a `Status` object instead of wrapping in an `error` object?

i.e.
```
{
    "code": ..
    "message":..
    "stack-trace":...
    "details":...
}
```

instead of

```
{
    "error": {
        "code": ..
        "message":..
        "stack-trace":...
        "details":...
    }
}
```

Reply:
Hi, wrap the status object in an error object is what google's error model does. ❤️

Reply:
I see 😅   I missed this the last time I reviewed, but other open source implementations don't seem to be wrapping in an `error` object.

https://github.com/line/armeria/pull/4231#discussion_r899950154

Reply:
Ah, I see. That's a decision we should make. I guess the purpose of wrapping an error object is making the error handling more extensive.

Reply:
Sounds reasonable 👍

Reply:
And also for client developers, when the error is wrapped in a single error object, it seems easier and more intuitive to detect the error message instead of having a flat normal response.

Reply:
Google's API Platform uses `error` object for backward compatibiltiy.
https://cloud.google.com/apis/design/errors#http_mapping